File: C-fast_integer/fft.c
#include <math.h>
#include "fft.h"

// Internal variables
static int phasevec_exist = 0;
static struct _sample phasevec[32];

// Public function
void fft(int log2point, struct _sample *restrict xy_out, const struct _sample *restrict xy_in) {
    int i;
    if (!phasevec_exist) {
	for (i=0; i<32; i++) {
	    int point = 2<<i;
	    phasevec[i].i = (1UL<<INTMUL)*cos(-2*M_PI/point);
	    phasevec[i].q = (1UL<<INTMUL)*sin(-2*M_PI/point);
	}
	phasevec_exist = 1;
    }
    for (i=0; i < (1<<log2point); i++) {
	unsigned int brev = i;
	brev = ((brev & 0xaaaaaaaa) >> 1) | ((brev & 0x55555555) << 1);
	brev = ((brev & 0xcccccccc) >> 2) | ((brev & 0x33333333) << 2);
	brev = ((brev & 0xf0f0f0f0) >> 4) | ((brev & 0x0f0f0f0f) << 4);
	brev = ((brev & 0xff00ff00) >> 8) | ((brev & 0x00ff00ff) << 8);
	brev = (brev >> 16) | (brev << 16);

	brev >>= 32-log2point;
	xy_out[brev] = xy_in[i];
    }

    // here begins the Danielson-Lanczos section
    int n = 1<<log2point;
    int l2pt=0;
    int mmax=1;
#ifdef MOD_SPEED
    while (l2pt < log2point) {
	int istep = 2<<l2pt;
#else
    while (n>mmax) {
	int istep = mmax<<1;
#endif
//	int theta = -2*M_PI/istep;
//	double complex wphase_XY = cos(theta) + sin(theta)*I;
	struct _sample wphase_XY = phasevec[l2pt];
	l2pt++;

	struct _sample w_XY = { 1UL<<INTMUL, 0 };
	for (int m=0; m < mmax; m++) {
	    for (int i=m; i < n; i += istep) {
		struct _sample tempXY;
		tempXY.i = ((MULTYPE)w_XY.i * xy_out[i+mmax].i - (MULTYPE)w_XY.q * xy_out[i+mmax].q) >> INTMUL;
		tempXY.q = ((MULTYPE)w_XY.i * xy_out[i+mmax].q + (MULTYPE)w_XY.q * xy_out[i+mmax].i) >> INTMUL;
		xy_out[i+mmax].i  = xy_out[i].i - tempXY.i;
		xy_out[i+mmax].q  = xy_out[i].q - tempXY.q;
		xy_out[i     ].i += tempXY.i;
		xy_out[i     ].q += tempXY.q;
	    }
	    struct _sample temp_w;
	    temp_w.i = ((MULTYPE)w_XY.i * wphase_XY.i - (MULTYPE)w_XY.q * wphase_XY.q) >> INTMUL;
	    w_XY.q = ((MULTYPE)w_XY.i * wphase_XY.q + (MULTYPE)w_XY.q * wphase_XY.i) >> INTMUL;
	    w_XY.i = temp_w.i;
	}
	mmax=istep;
    }
}

File: other_tests/Rust-tests/Rust-modular-safe-complex/src/fft.rs
extern crate num_complex;
use num_complex::Complex;

pub struct Fft {
    phasevec: [Complex<f64>; 32]
}

// Public function
impl Fft {
    pub fn new() -> Fft {
	let mut pvec: [Complex<f64>; 32] = [Complex::new(0.0, 0.0); 32];
	for i in 0..32 {
	    let point: i32 = 2<<i;
	    pvec[i] = Complex::new(
		(-2.0*std::f64::consts::PI/(point as f64)).cos(),
		(-2.0*std::f64::consts::PI/(point as f64)).sin()
	    );
	}
	Fft { phasevec: pvec}
    }

    pub fn fft(&self, log2point: u32, xy_out: &mut [Complex<f64>], xy_in: &[Complex<f64>]) {
	for i in 0 .. 1<<log2point {
	    let mut brev: usize = i;
	    brev = ((brev & 0xaaaaaaaa) >> 1) | ((brev & 0x55555555) << 1);
	    brev = ((brev & 0xcccccccc) >> 2) | ((brev & 0x33333333) << 2);
	    brev = ((brev & 0xf0f0f0f0) >> 4) | ((brev & 0x0f0f0f0f) << 4);
	    brev = ((brev & 0xff00ff00) >> 8) | ((brev & 0x00ff00ff) << 8);
	    brev = (brev >> 16) | (brev << 16);

	    brev >>= 32-log2point;
	    xy_out[brev] = xy_in[i];
	}

	// here begins the Danielson-Lanczos section;
	let n: usize = 1<<log2point;
	let mut l2pt: usize = 0;
	let mut mmax: usize = 1;

	while n > mmax {
	    let istep: usize = mmax<<1;
	    let wphase_xy: Complex<f64>;
	    wphase_xy = self.phasevec[l2pt];
	    l2pt+=1;
	    let mut w_xy: Complex<f64> = Complex::new(1.0, 0.0);

	    for m in 0..mmax {;
		let mut i = m;
		while i < n {
		    let temp: Complex<f64> = w_xy * xy_out[i+mmax];
		    xy_out[i+mmax] = xy_out[i] - temp;
		    xy_out[i     ] += temp;
		    i += istep;
		}
		w_xy *= wphase_xy; // rotate
	    }
	    mmax=istep;
	}
    }
}
